File: 40-combination-sum-ii/combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        self.answer = []
        self.candidates = sorted(candidates)
        self.recursion(target, [], 0)
        return self.answer

    def recursion(self, target, curr, i):
        if target ==  0:
            self.answer.append(curr.copy())
            return
        elif target < 0:
            return
        elif i >= len(self.candidates):
            return
        
        curr.append(self.candidates[i])
        self.recursion(target - self.candidates[i], curr, i+1)
        curr.pop()

        while i + 1 < len(self.candidates) and self.candidates[i] == self.candidates[i+1]:
            i += 1
        i += 1
        self.recursion(target, curr, i)
              
        
            
        

File: 40-combination-sum-ii/combination-sum-ii.java
class Solution {
    List<List<Integer>> answers;
    int[] candidates;
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        answers = new ArrayList<>();
        Arrays.sort(candidates);
        this.candidates = candidates;
        recursion(0, target, new ArrayList<>());
        return answers;
    }
    public void recursion(int i, int target, List<Integer> curr) {
        if (target == 0) {
            answers.add(new ArrayList<>(curr));
            return;
        } else if (target < 0) {
            return;
        } else if (i == candidates.length) {
            return;
        } else {
            curr.add(candidates[i]);
            recursion(i+1, target - candidates[i], curr);
            curr.removeLast();
            while (i + 1 < candidates.length && candidates[i] == candidates[i+1]) {
                i++;
            }
            recursion(i+1, target, curr);
        }
    }
}